text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,361 @@
+forwards that side-steps the time bomb: Do not allow fields with drop glue.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Union Definition
+
+When defining a union, it is a hard error to use a field type that requires drop glue.
+Examples:
+```rust
+// Accepted
+union Example1<T> {
+    // `ManuallyDrop<T>` never has drop glue, even if `T` does.
+    f1: ManuallyDrop<T>,
+    // `RefCell<i32>` is a fully known type, and does not have drop glue.
+    f2: RefCell<i32>,
+}
+union Example2<T: Copy> {
+    // `Copy` types never have drop glue.
+    f1: T,
+}
+trait Trait3 { type Assoc: Copy; }
+union Example3<T: Trait3> {
+    // `T::Assoc` is `Copy` and hence cannot have drop glue.
+    f1: T::Assoc,
+}
+
+// Rejected
+union Example4<T> {
+    // `T` might have drop glue, and then `Cell<T>` would as well.
+    f1: RefCell<T>,
+}
+trait Trait5 { type Assoc; }
+union Example5<T: Trait5> {
+    // `T::Assoc` might have drop glue.
+    f1: T::Assoc,
+}
+```
+
+Ruling out possibly dropping types may seem restrictive, but thanks to
+`ManuallyDrop` it in fact is not: If the compiler rejects a union definition,
+you can always wrap field types in `ManuallyDrop` to obtain a working
+definition.  This means you have to manually take care of when to drop the data,
+but that is already something to be concerned with when working on unions.
+
+As a consequence, it is quite obvious that writing to a union field will never
+implicitly call `drop`.  Such a write is hence always a safe operation.  This
+removes a whole class of pitfalls related to `drop` being called in tricky
+unsafe code when you might not expect that to happen.  (However, see below for
+some pitfalls that remain.)
+
+## Union initialization and `Drop`
+
+In two cases, the compiler cares about whether a (field of a) variable is
+initialized: When deciding whether a move from the field/variable is allowed
+(for cases where the type is not `Copy`), and when deciding whether or not the
+variable has to be dropped when it goes out of scope.
+
+A union just does very simple initialization tracking: There is a single boolean
+state for the entire union and all of its fields.  Nested inner fields are
+tracked just like they are for structs; however, when the union becomes
+(un)initialized, then all nested inner fields of all union fields are
+(un)initialized at once.  For example:
+
+```rust
+// This code creates bad references and transmutes to `Vec` in incorrect ways.
+// This is just to demonstrate what the compiler would accept in terms of
+// tracking initialization.
+
+struct S(i32); // not `Copy`, no drop glue
+union U { f1: ManuallyDrop<Vec<i32>>, f2: (S, S), f3: i32 }
+
+let mut u: U;
+// Now `u` is not initialized: `&u`, `&u.f2` and `&u.f2.0` are all rejected.
+
+// We can write into uninitialized inner fields:
+u.f2.1 = S(42);
+let _ = &u.f2.1; // This field is initialized now.
```

Comment:
`let _` is a bit of a footgun in terms of testing initialization.
`_` pattern doesn't access the right side *at all*, so even something like this
```rust
    let x: u8;
    let _ = x;
```
successfully compiles.
You need `let _y = x;` or just `x;` to test for `x` being initialized.

Reply:
*oops* thanks!